### PR TITLE
Some slicing fixes

### PR DESCRIPTION
**Pull request by mrocklin**
Using @shoyer 's slicing test from hell we've found a few meaningful bugs.  There are still a large number of failures coming from that test.  I'm not sure that I want to tackle them all.

See https://github.com/ContinuumIO/dask/pull/149

### CODE REVIEW

**Comment by mrocklin** [comment]
Thoughts on this?  Planning to merge.

**Comment by shoyer** [comment]
well, it's a step in the right direction :).

Did you figure out the case I mentioned in the original PR? `assert eq(x[:, [0, 1, 2]][[0, 1]], a[:, [0, 1, 2]][[0, 1]])`

I'm currently using indexing like that in xray routinely (to implement orthogonal indexing with dask arrays)... and it was working recently, before some of the latest slicing changes.

**Comment by mrocklin** [comment]
Fixed in https://github.com/ContinuumIO/dask/commit/97d91a84753bc5f5aab3ef7d0ec260b4dbd0cf0d

Sorry for glossing over that point.  PIcking out important failing use cases is very useful.

**Comment by mrocklin** [comment]
Further thoughts?

File: dask/array/tests/test_slicing.py [hunk 1]
```diff
@@ -378,3 +379,61 @@ def test_slice_list_then_None():
...
 
     assert eq(y, np.zeros((1, 2, 5)))
+
+
+class ReturnItem(object):
+    def __getitem__(self, key):
+        return key
+
+
+@skip
```
Comment (shoyer) [inline dask/array/tests/test_slicing.py]: possibly a better idea to use something like `unittest.skip` here? that way it gets listed as a skipped test and it's harder to forget about it :)

Comment (mrocklin) [inline dask/array/tests/test_slicing.py]: `unittest.skip` doesn't work on bare functions.  I could use either pytest.skip or nose.skip but then I have to choose.

Comment (shoyer) [inline dask/array/tests/test_slicing.py]: Hmm. That is slightly unfortunate.

**Comment by shoyer** [comment]
anyways, let's go ahead and merge, then.

**Pull request merged**